### PR TITLE
Add support for `IO::FileDescriptor` staying open on finalize

### PR DESCRIPTION
**Pull request by refi64**

### CODE REVIEW

File: src/io/file_descriptor.cr [hunk 1]
```diff
@@ -11,9 +11,13 @@ class IO::FileDescriptor < IO
     @volatile_fd.get
   end
 
-  def initialize(fd, blocking = nil)
+  # Whether or not to close the file descriptor when this object is finalized.
+  property close_on_finalize
+
+  def initialize(fd, blocking = nil, close_on_finalize = true)
```
Comment (straight-shoota) [inline src/io/file_descriptor.cr:17]: I think `close_on_finalize` should be a named argument.
`blocking` as well but changing that would be a breaking change (we could deprecated it, though. But that would be a different change).
When introducing a new parameter, we should make it a named parameter right away.
Adding a type restriction is also helpful to describe the API in detail.

```suggestion
  def initialize(fd, blocking = nil, *, close_on_finalize : Bool = true)
```

**Review by straight-shoota — COMMENTED** [review]
LGTM, except for two minor suggestions for improving the API.

**Comment by asterite** [comment]
Was there a discussion around this, or a reason to add it?

**Comment by refi64** [comment]
> Was there a discussion around this, or a reason to add it?

Welp the "throw out a PR right before night and leave the description blank" thing has come to bite me (again).

Essentially, when dealing with C APIs that return FDs that are owned by the callee, the only way to use `IO::FileDescriptor`'s higher-level API functions with it is by `dup`-ing it. That's...an entirely workable solution, but can come with its own share of annoyances (dup-ing a pipe and then accidentally leaving one end open). Of course, *not* dup-ing the fd can lead to "oh the C API closed it but you didn't notice"...but stuff like that is inherent to using both C APIs and raw FDs anyway.

I could certainly add this in local projects, but monkeypatching the stdlib isn't my favorite thing :sweat_smile:

**Comment by asterite** [comment]
I see, that makes sense! I would then add all of that description to the documentation of this new property. Explaining a concrete example behind this property will help users know when to use it, and know that they probably shouldn't use it at all 😅

**Comment by HertzDevil** [comment]
Related: https://github.com/crystal-lang/crystal/pull/12457#issuecomment-1239903346

File: src/io/file_descriptor.cr [hunk 2]
```diff
@@ -11,9 +11,13 @@ class IO::FileDescriptor < IO
     @volatile_fd.get
   end
 
-  def initialize(fd, blocking = nil)
+  # Whether or not to close the file descriptor when this object is finalized.
+  property close_on_finalize
```
Comment (HertzDevil) [inline src/io/file_descriptor.cr:15]: `#sync_close?` is a question getter, so we should follow suit:

```suggestion
  property? close_on_finalize : Bool
```

**Comment by refi64** [comment]
@asterite more detailed docs added

**Review by straight-shoota — APPROVED** [review]

**Pull request merged**